File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int dx[] = {0, 0, -1, 1}, dy[] = {1, -1, 0, 0};
int houseNum, cnt;
string house[25];

void dfs(int _x, int _y)
{
    if (_x == houseNum || _y == houseNum || _x < 0 || _y < 0)
        return;
    if (house[_y][_x] == '0')
        return;
    house[_y][_x] = '0';
    ++cnt;

    for (int i = 0; i < 4; i++)
        dfs(_x + dx[i], _y + dy[i]);
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);

    vector<int> rst;
    rst.reserve(313);
    cin >> houseNum;

    for (int i = 0; i < houseNum; i++)
        cin >> house[i];

    for (int i = 0; i < houseNum; i++)
    {
        for (int j = 0; j < houseNum; j++)
        {
            if (house[i][j] == '1')
            {
                cnt = 0;
                dfs(j, i);
                rst.emplace_back(cnt);
            }
        }
    }

    sort(rst.begin(), rst.end());
    
    cout << rst.size() << '\n';
    for (auto &a : rst)
        cout << a << '\n';

    return 0;
}

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.io.*;
import java.util.*;

public class Main {
    static int[] dx = {0, 0, -1, 1};
    static int[] dy = {1, -1, 0, 0};
    static int houseNum;
    static int cnt;
    static char[][] house;

    static void dfs(int x, int y) {
        if (x < 0 || y < 0 || x >= houseNum || y >= houseNum || house[y][x] == '0') {
            return;
        }
        house[y][x] = '0';
        cnt++;

        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            dfs(nx, ny);
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        houseNum = Integer.parseInt(st.nextToken());
        house = new char[houseNum][houseNum];

        for (int i = 0; i < houseNum; i++) {
            house[i] = br.readLine().toCharArray();
        }

        ArrayList<Integer> rst = new ArrayList<>();
        for (int i = 0; i < houseNum; i++) {
            for (int j = 0; j < houseNum; j++) {
                if (house[i][j] == '1') {
                    cnt = 0;
                    dfs(j, i);
                    rst.add(cnt);
                }
            }
        }

        Collections.sort(rst);
        System.out.println(rst.size());
        for (int num : rst) {
            System.out.println(num);
        }
        br.close();
    }
}